File: DP/2667.py
from collections import deque

def bfs(graph, i, j):
  q = deque()
  q.append((i, j))
  count = 1
  graph[i][j] = 0
  while q:
    x, y = q.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]
      if nx < 0 or ny < 0 or nx >= n or ny >= n:
        continue
      if graph[nx][ny] == 1:
        q.append((nx, ny))
        graph[nx][ny] = 0
        count += 1
  ans.append(count)

n = int(input())
graph = []
ans = []
for _ in range(n):
  graph.append(list(map(int, input())))
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

count = 0
for i in range(n):
  for j in range(n):
    if graph[i][j] == 1:
      bfs(graph, i, j)
      count += 1

print(count)
ans.sort()

for a in ans:
  print(a)

File: DP/2667.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <bits/stdc++.h>
#define MAX 26

using namespace std;

int graph[MAX][MAX];
int dx[] = { -1, 1, 0, 0 };
int dy[] = { 0, 0, -1, 1 };
bool visited[MAX][MAX];
vector<int> vec;

void bfs(int xPos, int yPos, int n) {
	queue<pair<int, int>> q;
	q.push({ xPos, yPos });
	int count = 1;
	visited[xPos][yPos] = true;
	while (!q.empty()) {
		int x = q.front().first;
		int y = q.front().second;
		q.pop();
		for (int i = 0; i < 4; i++) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			if (nx < 0 || ny < 0 || nx >= n or ny >= n)
				continue;
			if (graph[nx][ny] != 0 && visited[nx][ny] == false) {
				q.push({ nx, ny });
				count += 1;
				visited[nx][ny] = true;
			}
		}
	}
	vec.push_back(count);
}

int main(void) {
	int n;
	scanf("%d", &n);

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			scanf("%1d", &graph[i][j]);
		}
	}

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (graph[i][j] != 0 && visited[i][j] == false) {
				bfs(i, j, n);
			}
		}
	}

	sort(vec.begin(), vec.end());

	printf("%d\n", vec.size());
	for (int i = 0; i < vec.size(); i++) {
		printf("%d\n", vec[i]);
	}
	return 0;
}
